File: clients/rn/__tests__/engine/room-memory-policy.test.ts
import { RoomMemoryPolicy } from '../../src/engine/room/RoomMemoryPolicy';
import type { Said } from '../../src/engine/events/WorldEvent';

function makeSaid(text: string): Said {
  return {
    type: 'said', roomId: 'test', timestamp: Date.now(),
    entityId: 'p1', entityName: 'Alice', text,
  };
}

describe('RoomMemoryPolicy', () => {
  it('adds events to hot buffer', () => {
    const policy = new RoomMemoryPolicy(10, 20);
    policy.add(makeSaid('Hello'));
    expect(policy.getHotEvents()).toHaveLength(1);
    expect(policy.getWarmEvents()).toHaveLength(0);
  });

  it('overflows hot to warm when hot exceeds size', () => {
    const policy = new RoomMemoryPolicy(3, 20);
    for (let i = 0; i < 5; i++) {
      policy.add(makeSaid(`msg-${i}`));
    }
    expect(policy.getHotEvents()).toHaveLength(3);
    expect(policy.getWarmEvents()).toHaveLength(2);
    expect(policy.getHotEvents()[0].text).toBe('msg-2');
    expect(policy.getWarmEvents()[0].text).toBe('msg-0');
  });

  it('drops warm events when warm exceeds size', () => {
    const policy = new RoomMemoryPolicy(2, 3);
    for (let i = 0; i < 10; i++) {
      policy.add(makeSaid(`msg-${i}`));
    }
    expect(policy.getHotEvents()).toHaveLength(2);
    expect(policy.getWarmEvents()).toHaveLength(3);
    // Oldest warm events were dropped
    expect(policy.getAllEvents()).toHaveLength(5);
  });

  it('clear resets all buffers', () => {
    const policy = new RoomMemoryPolicy(5, 5);
    for (let i = 0; i < 8; i++) {
      policy.add(makeSaid(`msg-${i}`));
    }
    policy.clear();
    expect(policy.getHotEvents()).toHaveLength(0);
    expect(policy.getWarmEvents()).toHaveLength(0);
    expect(policy.getAllEvents()).toHaveLength(0);
  });
});

File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/room/RoomMemoryPolicyTest.kt
package org.wyrdsekai.app.engine.room

import kotlin.time.Clock
import org.wyrdsekai.app.engine.event.WorldEvent
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class RoomMemoryPolicyTest {

    private fun makeSaid(text: String) = WorldEvent.Said(
        roomId = "test",
        timestamp = Clock.System.now(),
        entityId = "player-1",
        entityName = "Alice",
        text = text,
    )

    @Test
    fun defaultPolicyHasCorrectSizes() {
        val policy = RoomMemoryPolicy.default()
        // Default hot=10, warm=20 — verify by filling up to boundary
        // Add 10 events: all should be in hot
        repeat(10) { policy.add(makeSaid("msg-$it")) }
        assertEquals(10, policy.hotEvents().size)
        assertEquals(0, policy.warmEvents().size)

        // 11th event should push one to warm
        policy.add(makeSaid("msg-10"))
        assertEquals(10, policy.hotEvents().size)
        assertEquals(1, policy.warmEvents().size)
    }

    @Test
    fun addToHotBuffer() {
        val policy = RoomMemoryPolicy.default()
        val event = makeSaid("Hello")
        policy.add(event)
        assertEquals(1, policy.hotEvents().size)
        assertEquals("Hello", policy.hotEvents().first().text)
    }

    @Test
    fun hotOverflowsToWarm() {
        val policy = RoomMemoryPolicy(hotSize = 3, warmSize = 5)

        repeat(5) { policy.add(makeSaid("msg-$it")) }

        // Hot should have the last 3
        assertEquals(3, policy.hotEvents().size)
        assertEquals("msg-2", policy.hotEvents()[0].text)
        assertEquals("msg-3", policy.hotEvents()[1].text)
        assertEquals("msg-4", policy.hotEvents()[2].text)

        // Warm should have the first 2 that overflowed
        assertEquals(2, policy.warmEvents().size)
        assertEquals("msg-0", policy.warmEvents()[0].text)
        assertEquals("msg-1", policy.warmEvents()[1].text)
    }

    @Test
    fun getHotEventsReturnsCorrectCount() {
        val policy = RoomMemoryPolicy.default()
        policy.add(makeSaid("one"))
        policy.add(makeSaid("two"))
        policy.add(makeSaid("three"))
        assertEquals(3, policy.hotEvents().size)
    }

    @Test
    fun clearResetsBuffers() {
        val policy = RoomMemoryPolicy(hotSize = 2, warmSize = 2)
        repeat(4) { policy.add(makeSaid("msg-$it")) }

        assertTrue(policy.hotEvents().isNotEmpty())
        assertTrue(policy.warmEvents().isNotEmpty())

        policy.clear()

        assertTrue(policy.hotEvents().isEmpty())
        assertTrue(policy.warmEvents().isEmpty())
        assertTrue(policy.allEvents().isEmpty())
    }
}
